### PR TITLE
Harden Linux liveview stop teardown

### PR DESCRIPTION
**Pull request by thomashenauer**
# Description

In affected Linux wxMediaCtrl/GStreamer setups, stopping liveview through the legacy backend did **not** reliably close the underlying RTSP/RTSPS stream session. In practice, the transport could remain active after pressing Stop, leaving stale stream state and causing unstable restart behavior.

This PR hardens Linux liveview stop/teardown for the legacy wxMediaCtrl/GStreamer backend. The native Wayland GTK sink path is preserved.

## What this PR changes

- Linux GStreamer stop teardown in `wxMediaCtrl2::Stop()` for the legacy wxMediaCtrl backend:
  - set playbin to `GST_STATE_READY`
  - wait briefly for transition
  - set playbin to `GST_STATE_NULL`
  - wait for stop transition completion
  - clear playbin URI

## User-visible result

- Stopping liveview should more reliably tear down the RTSP/RTSPS stream session on Linux X11 / wxMediaCtrl fallback paths.
- Restart behavior is more robust when the previous stream did not shut down cleanly.

## Scope / non-goals

- Scope is limited to Linux liveview teardown in the legacy wxMediaCtrl/GStreamer backend.
- Native Wayland GTK sink behavior is unchanged.
- No `MediaPlayCtrl` UI state-machine changes in the final branch.
- No printer protocol changes.
- No preset/profile/translation changes.
- A non-fatal GStreamer warning (`gst_video_center_rect: assertion 'src->h != 0' failed`) may still appear in frozen H2D live view edge cases.

# Screenshots/Recordings/Graphs

N/A (lifecycle/state fix, no visual layout change).

## Tests

Date tested: May 16, 2026  
Platform: Arch Linux, GStreamer 1.28.1.

### Build verification

- Configured against current upstream dependency build with wxWidgets 3.3.2.
- `cmake --build build-pr12552 --config Release --target OrcaSlicer -j4` succeeded.

### Automated tests

- Not rerun after rebasing to current `main`.

### Manual verification

- Regression flow to test:
  - start liveview
  - stop liveview
  - re-enable liveview
  - repeat stop/play actions
- Result to verify:
  - RTSP/RTSPS stream connection is released after stop
  - restart does not reuse stale stream state
  - no crash observed

### Connection lifecycle check (`ss -tpn`)

- Before enabling liveview: MQTT (`:8883`) only
- While liveview enabled: MQTT (`:8883`) + RTSP (`:322`)
- After stop: RTSP (`:322`) disappears, MQTT remains

Tested printers:

- Bambu X1C
- Bambu H2D

### CODE REVIEW

File: src/slic3r/GUI/MediaPlayCtrl.cpp [hunk 1]
```diff
@@ -616,9 +619,11 @@ void MediaPlayCtrl::onStateChanged(wxMediaEvent &event)
         wxSize size = m_media_ctrl->GetVideoSize();
         BOOST_LOG_TRIVIAL(info) << "MediaPlayCtrl::onStateChanged: size: " << size.x << "x" << size.y;
         m_failed_code = m_media_ctrl->GetLastError();
+        const bool stop_requested = m_stop_requested.load(std::memory_order_relaxed);
         if (size.GetWidth() >= 320) {
             m_last_state = state;
             m_failed_code = 0;
+            m_stop_requested = false;
```
[the remark this answers is not in the document]
Comment (thomashenauer) [reply src/slic3r/GUI/MediaPlayCtrl.cpp]: Good point. This flag only needs atomicity here, not stronger cross-variable ordering, so I changed the stores to explicit `store(..., std::memory_order_relaxed)` to match the existing relaxed load. Addressed in `ecb0f8a3bf`

File: src/slic3r/GUI/wxMediaCtrl2.cpp [hunk 2]
```diff
@@ -213,6 +213,19 @@ void wxMediaCtrl2::Play() { wxMediaCtrl::Play(); }
...
 {
+#ifdef __LINUX__
+    m_loaded = false;
+    if (m_imp != nullptr) {
+        auto playbin = reinterpret_cast<wxGStreamerMediaBackend *>(m_imp)->m_playbin;
+        if (playbin) {
+            gst_element_set_state(playbin, GST_STATE_READY);
+            gst_element_get_state(playbin, nullptr, nullptr, 200 * GST_MSECOND);
+            gst_element_set_state(playbin, GST_STATE_NULL);
+            gst_element_get_state(playbin, nullptr, nullptr, 1000 * GST_MSECOND);
```
[the remark this answers is not in the document]
Comment (thomashenauer) [reply src/slic3r/GUI/wxMediaCtrl2.cpp]: Agreed. I added explicit `GstStateChangeReturn` handling and warning logs around the READY and NULL teardown transitions, and I avoid waiting when `gst_element_set_state()` fails immediately. Addressed in `ecb0f8a3bf`.

File: src/slic3r/GUI/wxMediaCtrl2.cpp [hunk 3]
```diff
@@ -213,6 +213,19 @@ void wxMediaCtrl2::Play() { wxMediaCtrl::Play(); }
...
+            g_object_set(G_OBJECT(playbin), "uri", NULL, NULL);
```
[the remark this answers is not in the document]
Comment (thomashenauer) [reply src/slic3r/GUI/wxMediaCtrl2.cpp]: I switched that `g_object_set(..., "uri", ...)` call to `nullptr`. Addressed in `ecb0f8a3bf`.

**Comment by SoftFever** [comment]
Hi @thomashenauer 
Thanks for the work.
Let me know if you need help for the test.

**Comment by thomashenauer** [comment]
Thanks @SoftFever 
This PR is from before v2.3.2 when I was debugging the freezing liveview issue that was later fixed by #13359.
With the v2.3.2 release I'm still able the reproduce stale RTSP streams (port 322) after the liveview stream frezzes or is stopped (except I can't really stop before it freezes or the UI hangs).
With the current main build the liveview and device tab is much improved and I can't produce a stale RTSP stream on Wayland or X11. 
On Wayland it makes sense because we switched to a GTK sink path away from the wxMediaCtrl/GStreamer path (#13432). 
On X11, current main still keeps the old pipeline which caused the stale streams on v2.3.2 but wx was updated and I can't reproduce the stale streams even on X11 anymore. 

Maybe you could test for me if the RTSP stream for bambu liveviews are correctly terminating on your system. 
I used 
``` pid=$(pgrep -n -f 'OrcaSlicer|orca-slicer') ```
``` watch -n 0.2 "ss -Htanp | rg 'printer_ip|pid=$pid,'" ```
to monitor the connection to my Lan only X1C and H2D.

MQTT (:8883) - always active when connected in device tap
RTSP (:322) - should only be active when liveview is active.

I couldn't find evidence for what exactly fixed the issue for me in current main on X11. wx’s Linux GStreamer Stop seems to still not tear down the playbin. Might be that the stale streams where a result of the stream freezes first, and then Stop is operating against a stuck/odd pipeline state.

This PR could still have a validity as a robustness fix.

We could reframe/rework the fix as “makes Linux wxGStreamer liveview stop deterministic by forcing the underlying playbin to NULL and clearing the URI, covering cases where wxMediaCtrl::Stop only pauses/seeks.”

Would you accept a narrow defensive teardown, or should we close this unless someone can reproduce stale sockets on main?

I'm currently building a follow up commit that resolves the current conflicts and scopes down to a purely defensive robusness fix for the wxMediaCtrl/GStreamer backend. I'll push once I confirmed that it does not exhibit regressions.

**Comment by SoftFever** [comment]
> Thanks @SoftFever This PR is from before v2.3.2 when I was debugging the freezing liveview issue that was later fixed by #13359. With the v2.3.2 release I'm still able the reproduce stale RTSP streams (port 322) after the liveview stream frezzes or is stopped (except I can't really stop before it freezes or the UI hangs). With the current main build the liveview and device tab is much improved and I can't produce a stale RTSP stream on Wayland or X11. On Wayland it makes sense because we switched to a GTK sink path away from the wxMediaCtrl/GStreamer path (#13432). On X11, current main still keeps the old pipeline which caused the stale streams on v2.3.2 but wx was updated and I can't reproduce the stale streams even on X11 anymore.
> 
> Maybe you could test for me if the RTSP stream for bambu liveviews are correctly terminating on your system. I used `pid=$(pgrep -n -f 'OrcaSlicer|orca-slicer')` `watch -n 0.2 "ss -Htanp | rg 'printer_ip|pid=$pid,'"` to monitor the connection to my Lan only X1C and H2D.
> 
> MQTT (:8883) - always active when connected in device tap RTSP (:322) - should only be active when liveview is active.
> 
> I couldn't find evidence for what exactly fixed the issue for me in current main on X11. wx’s Linux GStreamer Stop seems to still not tear down the playbin. Might be that the stale streams where a result of the stream freezes first, and then Stop is operating against a stuck/odd pipeline state.
> 
> This PR could still have a validity as a robustness fix.
> 
> We could reframe/rework the fix as “makes Linux wxGStreamer liveview stop deterministic by forcing the underlying playbin to NULL and clearing the URI, covering cases where wxMediaCtrl::Stop only pauses/seeks.”
> 
> Would you accept a narrow defensive teardown, or should we close this unless someone can reproduce stale sockets on main?
> 
> I'm currently building a follow up commit that resolves the current conflicts and scopes down to a purely defensive robusness fix for the wxMediaCtrl/GStreamer backend. I'll push once I confirmed that it does not exhibit regressions.

Tested with P1S and H2D, I can't repro this issue too.
Looks like fixed now

**Comment by thomashenauer** [comment]
@SoftFever 
The error surface seems to be mostly theoretical now or at least quite small and even then limited to X11. 
Current change is already scoped back to just a hardening fix for the remaining potential edge cases.
Since the failure mode ist just a stale RTSP stream, I would probably tend to dismiss this fix as obsolete/won't fix.
If you agree you can close this PR yourself or tell me to do so.

**Comment by SoftFever** [comment]
> @SoftFever 
> 
> The error surface seems to be mostly theoretical now or at least quite small and even then limited to X11. 
> 
> Current change is already scoped back to just a hardening fix for the remaining potential edge cases.
> 
> Since the failure mode ist just a stale RTSP stream, I would probably tend to dismiss this fix as obsolete/won't fix.
> 
> If you agree you can close this PR yourself or tell me to do so. 

Agreed.👍
